### PR TITLE
offset: allow zero-byte offset on arbitrary pointers

### LINKED ISSUE #65108: ptr::offset should explicitly clarify 0-sized offset semantics
https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

I can't remember the past arguments we definitely had about this, but it would be nice to explicitly call out the answer in the docs.

As a relevant example, Vec::into_iter currently unconditionally computes the "end" pointer of the array by offsetting by `len` (as long as size_of T > 0). This means that we offset a dangling pointer by 0 when iterating an empty Vec. This is obviously useful to support and annoying to have to guard against, so I would hope that's well-defined.

https://doc.rust-lang.org/src/alloc/vec.rs.html#1860

```rust
fn into_iter(mut self) -> IntoIter<T> {
        unsafe {
            let begin = self.as_mut_ptr();
            let end = if mem::size_of::<T>() == 0 {
                arith_offset(begin as *const i8, self.len() as isize) as *const T
            } else {
                // SAFE when `begin` dangles and `len == 0`???
                begin.add(self.len()) as *const T
            };
            let cap = self.buf.capacity();
            mem::forget(self);
            IntoIter {
                buf: NonNull::new_unchecked(begin),
                phantom: PhantomData,
                cap,
                ptr: begin,
                end,
            }
        }
    }
```

### PR DESCRIPTION
**Pull request by RalfJung**
As per prior @rust-lang/opsem [discussion](https://github.com/rust-lang/opsem-team/issues/10) and [FCP](https://github.com/rust-lang/unsafe-code-guidelines/issues/472#issuecomment-1793409130):

- Zero-sized reads and writes are allowed on all sufficiently aligned pointers, including the null pointer
- Inbounds-offset-by-zero is allowed on all pointers, including the null pointer
- `offset_from` on two pointers derived from the same allocation is always allowed when they have the same address

This removes surprising UB (in particular, even C++ allows "nullptr + 0", which we currently disallow), and it brings us one step closer to an important theoretical property for our semantics ("provenance monotonicity": if operations are valid on bytes without provenance, then adding provenance can't make them invalid).

The minimum LLVM we require (v17) includes https://reviews.llvm.org/D154051, so we can finally implement this. 

The `offset_from` change is needed to maintain the equivalence with `offset`: if `let ptr2 = ptr1.offset(N)` is well-defined, then `ptr2.offset_from(ptr1)` should be well-defined and return N. Now consider the case where N is 0 and `ptr1` dangles: we want to still allow offset_from here.

I think we should change offset_from further, but that's a separate discussion.

Fixes https://github.com/rust-lang/rust/issues/65108
[Tracking issue](https://github.com/rust-lang/rust/issues/117945) | [T-lang summary](https://github.com/rust-lang/rust/pull/117329#issuecomment-1951981106)

Cc @nikic

### CODE REVIEW

**Comment by RalfJung** [comment]
Uh, what...? Removing the `inbounds` leads to a segfault...?

**Comment by nikic** [comment]
Dropping inbounds changes optimization behavior in a really major way, which is likely to expose latent issues. Keeping inbounds for all LLVM versions is a lot less risky.

**Comment by RalfJung** [comment]
@nikic  That's fine for me, I just wonder whether it can cause issues on its own when people start relying on `offset(0)` being allowed. LLVM was changed in two places to make that legal:
- https://reviews.llvm.org/D154054
- https://reviews.llvm.org/D154215

Could one write Rust code that is legal under the new model but miscompiled by old versions of LLVM?

How fixed is our LLVM support policy -- is it always the last 3 versions, or once LLVM 18 is released, could we say that we require LLVM 17?

**Comment by nikic** [comment]
> Could one write Rust code that is legal under the new model but miscompiled by old versions of LLVM?

I think the second patch is not relevant to rust. The first one could in theory be, but you'd need some quite exotic, artificially constructed code for it. I don't think there is any chance that this would affect "real" rust code.

> How fixed is our LLVM support policy -- is it always the last 3 versions, or once LLVM 18 is released, could we say that we require LLVM 17?

The earliest we can drop LLVM 16 support is when Fedora 38 goes EOL, which would be in May.

**Comment by apiraino** [comment]
If I read the prev comment correctly, this PR can switch back to the author.

@rustbot author

**Comment by RalfJung** [comment]
Nominating for t-lang discussion. This implements the t-opsem consensus from https://github.com/rust-lang/opsem-team/issues/10, https://github.com/rust-lang/unsafe-code-guidelines/issues/472 to generally allow zero-sized accesses on all pointers. Also see the [tracking issue](https://github.com/rust-lang/rust/issues/117945).

- Zero-sized reads and writes are allowed on all sufficiently aligned pointers, including the null pointer
- Inbounds-offset-by-zero is allowed on all pointers, including the null pointer
- `offset_from` on two pointers is always allowed when they have the same address (but see the caveat below)

This means the following function is safe to be called on any pointer:
```rust
fn test_ptr(ptr: *mut ()) { unsafe {
    // Reads and writes.
    let mut val = *ptr;
    *ptr = val;
    ptr.read();
    ptr.write(());
    // Memory access intrinsics.
    // - memcpy (1st and 2nd argument)
    ptr.copy_from_nonoverlapping(&(), 1);
    ptr.copy_to_nonoverlapping(&mut val, 1);
    // - memmove (1st and 2nd argument)
    ptr.copy_from(&(), 1);
    ptr.copy_to(&mut val, 1);
    // - memset
    ptr.write_bytes(0u8, 1);
    // Offset.
    let _ = ptr.offset(0);
    let _ = ptr.offset(1); // this is still 0 bytes
    // Distance.
    let ptr = ptr.cast::<i32>();
    ptr.offset_from(ptr);
} }
```

Some specific concerns warrant closer scrutiny.

## LLVM 16

We currently still support LLVM 16, which does not yet have the patches that make `getelementptr inbounds` always well-defined for offset 0. The function above thus generates LLVM IR with UB. No known miscompilations arise from that, and my attempt at just removing the `inbounds` annotation on old versions of LLVM failed (I got segfaults, and Nikic [suggested](https://github.com/rust-lang/rust/pull/117329#issuecomment-1783925317) that keeping these attribute around is actually less risky than removing them). If we want to avoid this, we have to wait until support for LLVM 16 can be dropped (which apparently is in May).

## Null pointers

t-opsem decided to allow zero-sized reads and writes on null pointers. This is mostly for consistency: we definitely want to allow zero-sized offsets on null pointers (`ptr::null::<T>().offset(0)`), since this is allowed in C++ (and a proposal is being made to allow it in C) and there's no reason for us to have more UB than C++ here. But if we allow this, and therefore consider the null pointer to have a zero-sized region of "inbounds" memory, then it would be inconsistent to not allow reading from / writing to that region.

## `offset_from`

This operation is somewhat special as it takes two pointers. We do want `test_ptr` above to be defined on all pointers, so `offset_from` between two identical pointers without provenance must be allowed. But we also want to achieve this property called "provenance monotonicity", whereby adding arbitrary provenance to any no-provenance pointer must never make the program UB.[^mono] From these two it follows that calling `offset_from` with two pointers with the same address but arbitrary different provenance must be allowed. This does have some [minor downsides](https://github.com/rust-lang/unsafe-code-guidelines/issues/472#issuecomment-1921686682). So my proposal (and this goes beyond what t-opsem agreed on) is to define the `ptr_offset_from` intrinsic to satisfy provenance monotonicity, but to document the user-facing `ptr.offset_from(...)` as requiring either two pointers without provenance or two pointers with provenance for the same allocation -- therefore, making the case of provenance mismatch library UB, but not language UB.

[^mono]: This property should hopefully make some intuitive sense, and it is also crucial to justify optimizations that make the program have more provenance than before -- such as optimizing away provenance-stripping operations. Specifically, `*ptr = *ptr` where `ptr: *mut usize` is likely going to be a provenance-stripping operation, and so optimizing away this redundant assignment requires provenance monotonicity.

File: src/tools/miri/tests/pass/zero-sized-accesses-and-offsets.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        ptr.write_bytes(0u8, 1);
+        // Offset.
+        let _ = ptr.offset(0);
+        let _ = ptr.offset(1); // this is still 0 bytes
+        // Distance.
+        let ptr = ptr.cast::<i32>();
+        ptr.offset_from(ptr);
+        // Distance from other "bad" pointers that have the same address, but different provenance. Some
+        // of this is library UB, but we don't want it to be language UB since that would violate
+        // provenance monotonicity: if we allow computing the diststance between two ptrs with no
```
Comment (oli-obk) [inline src/tools/miri/tests/pass/zero-sized-accesses-and-offsets.rs:44]:
```suggestion
        // provenance monotonicity: if we allow computing the distance between two ptrs with no
```

File: tests/ui/consts/offset_from_ub.rs [hunk 2]
```diff
@@ -122,4 +108,16 @@ pub const OFFSET_VERY_FAR2: isize = {
     //~^ inside
 };
 
+// If the pointers are the same, OOB/null is fine.
```
Comment (oli-obk) [inline tests/ui/consts/offset_from_ub.rs:111]: test use after-free, too

**Review by oli-obk — APPROVED** [review]
impl lgtm

File: src/tools/miri/tests/pass/zero-sized-accesses-and-offsets.rs (hunk 1 above)
Comment (Dante-Broggi) [inline src/tools/miri/tests/pass/zero-sized-accesses-and-offsets.rs:44]:
```suggestion
        // provenance monotonicity: if we allow computing the distance between two ptrs with no
```

**Comment by traviscross** [comment]
We discussed this in lang triage today.  That discussion led to a number of clarification and other questions that @scottmcm, @tmandry, and @pnkfelix will be asking here.

**Comment by tmandry** [comment]
> So my proposal (and this goes beyond what t-opsem agreed on) is to define the `ptr_offset_from` intrinsic to satisfy provenance monotonicity, but to document the user-facing `ptr.offset_from(...)` as requiring either two pointers without provenance or two pointers with provenance for the same allocation -- therefore, making the case of provenance mismatch library UB, but not language UB. This gives us some wiggle room to course-correct later if we find a better option.

So is it important to have _some_ operation that satisfies provenance monotonicity, but not that _all_ operations satisfy it? Why?

> Specifically, `*ptr = *ptr` where `ptr: *mut usize` is likely going to be a provenance-stripping operation, and so optimizing away this redundant assignment requires provenance monotonicity.

Can you elaborate on why this is a provenance stripping operation?

**Comment by RalfJung** [comment]
> So is it important to have some operation that satisfies provenance monotonicity, but not that all operations satisfy it? Why?

The entire operational semantics needs to satisfy provenance monotonicity. But the library can impose further UB that the compiler knows nothing about, and that UB does not have to satisfy provenance monotonicity.

>    Can you elaborate on why this is a provenance stripping operation?

Because integers cannot carry provenance, so in `*ptr = *ptr`, the LHS evaluate to a value that does not have provenance, and then that gets stored to memory. Thus after that statement, the memory behind `*ptr` definitely has no provenance any more, even if it did have provenance before.

**Comment by RalfJung** [comment]
I have changed the PR to remove the controversial/confusing offset_from parts. There are still some changes in offset_from behavior, but they are hopefully entirely non-surprising. The [nomination comment](https://github.com/rust-lang/rust/pull/117329#issuecomment-1951981106) has been updated.

**Comment by RalfJung** [comment]
@nikic just to be sure -- this does not require any LLVM changes, right?

- allow `offset(0)` on all pointers
- allow `ptr_offset_from` on two pointers derived from the same allocation if they have the same address, even if they are out-of-bounds or the allocation is dead
- allow zero-sized read and write accesses on all pointers. For native reads/writes via `*` and `=`, we don't even emit any LLVM IR when the size is 0 (the type `i0` does not even exist), so this only affects memcpy/memmove/memset.

We already confirmed cranelift doesn't need any changes.

@antoyo @GuillaumeGomez does the GCC backend need changes for this? If yes, probably there should be an issue opened somewhere to track that.
- If I understand [this](https://github.com/rust-lang/rust/blob/1ce5dc8d9c638194306e63abfe81748c77e7386e/compiler/rustc_codegen_gcc/src/builder.rs#L1166-L1182) correctly, the "inbounds" `offset` is translated to an array access. (That's a bit surprising, I would have expected a form of pointer arithmetic, like whatever `ptr + i` in C translates to.) So it is important that array accesses at offset 0 are permitted on arbitrary dangling or null pointers.
- For memcpy/memmove/memset, whatever GCC operations these translate to should allow arbitrary dangling pointers, including NULL, if the size is 0.

**Comment by antoyo** [comment]
We will probably need changes to the GCC backend and libgccjit as well to allow `+` on pointer types.

For `mem*`, it uses the GCC builtins, so it depends on their behavior.

**Comment by RalfJung** [comment]
Where's the right place to file an issue to track that?

(Given that the GCC backend is incomplete, I'm assuming this is not a blocker for this PR.)

**Comment by tmandry** [comment]
@rfcbot reviewed


> Because integers cannot carry provenance, so in `*ptr = *ptr`, the RHS (of type `usize`) evaluates to a value that does not have provenance, and then that gets stored to memory. Thus after that statement, the memory behind `*ptr` definitely has no provenance any more, even if it did have provenance before.

Yes, that makes sense – the thing that I missed or was unclear was that `ptr` is a pointer to a pointer in this case.

**Comment by antoyo** [comment]
> Where's the right place to file an issue to track that?
> 
> (Given that the GCC backend is incomplete, I'm assuming this is not a blocker for this PR.)

I would say either add a `TODO` in the code or open an issue [on the cg_gcc repo](https://github.com/rust-lang/rustc_codegen_gcc).

**Comment by RalfJung** [comment]
> Yes, that makes sense – the thing that I missed or was unclear was that ptr is a pointer to a pointer in this case.

It's a pointer to an *integer*. Specifically, the type of `ptr` is `*mut usize`.
But --  the data stored as `*ptr` is a pointer; I guess that's what you mean. Yes it would have been good to say that. :) I have updated the more extended explanation at https://github.com/rust-lang/rust/pull/124921 based on your feedback.

**Comment by RalfJung** [comment]
Okay, filed some issues for the GCC backend:
- https://github.com/rust-lang/rustc_codegen_gcc/issues/515
- https://github.com/rust-lang/rustc_codegen_gcc/issues/516

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by RalfJung** [comment]
>  My recollection of the room was that the .offset(0) parts had broad agreement, and sticking to just the "well we need offset and offset_from to be inverses" part there as you wrote it makes good sense to me. So let's try to land this part

@scottmcm  There's not just the `offset(0)` part, there's also the "zero-sized memory accesses" part. I just realized I forgot to mention that in the PR summary, but it *was* in the [nomination comment](https://github.com/rust-lang/rust/pull/117329#issuecomment-1951981106). Did that come up during t-lang discussion? If not we should probably restart FCP to make sure everyone is aware of all three things that are changed here:

- Zero-sized reads and writes are allowed on all sufficiently aligned pointers, including the null pointer
- Inbounds-offset-by-zero is allowed on all pointers, including the null pointer
- `offset_from` on two pointers derived from the same allocation is always allowed when they have the same address

**Comment by nikic** [comment]
> @nikic just to be sure -- this does not require any LLVM changes, right?
> 
>     * allow `offset(0)` on all pointers

This is fine.

>     * allow `ptr_offset_from` on two pointers derived from the same allocation if they have the same address, even if they are out-of-bounds or the allocation is dead

This is certainly fine right now, as we emit pointer subtractions in a way that does not impose any (LLVM) requirements on them.

If we introduce an intrinsic for this on the LLVM side, I'd expect that we would be able to justify having compatible semantics for the sake of symmetry with GEP inbounds.

>     * allow zero-sized read and write accesses on all pointers. For native reads/writes via `*` and `=`, we don't even emit any LLVM IR when the size is 0 (the type `i0` does not even exist), so this only affects memcpy/memmove/memset.

This is fine per LLVM LangRef. As you know, the situation wrt libc is a bit fuzzy, but hopefully https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3177.pdf will resolve this. If not, this means that rustc requires a compatible libc implementation.

**Comment by RalfJung** [comment]
We have an r=me from oli for the impl, but I did tweak the doc comments since then so would be good if someone could review those.

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 3]
```diff
@@ -434,6 +434,10 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
                 found_bytes: has.bytes()
             },
         );
+        // Make sure this is non-null. (ZST references can be dereferenceable and null.)
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/validity.rs:437]: This does refer to "`&`s".

But the comment isn't very clear, I will improve it.

**Pull request merged**